### DISCUSSION TITLE
Is this the correct way to set up a once daily cron job?

### DISCUSSION THREAD

**Discussion by assembly-winston** [riverqueue/river#521] [Q&A] [answered]
Hi River community,

I am in need of a cron job on my river queue that runs once daily. I do not want it to be affected by deploys (my worry is that since the following code is in main, a new cron job gets added every time I deploy the server, instead of just running regardless).

```
        w := river.NewWorkers()
	river.AddWorker(w, &CronWorker{DB: storage, Logger: logger})

        config = &river.Config{
	        Queues: map[string]river.QueueConfig{
		        river.QueueDefault: {MaxWorkers: 100},
	        },
	        Workers: w,
        }
		
	cli, err := river.NewClient(
		riverpgxv5.New(pool),
		config,
	)
        if err != nil {
		return nil, false, err
	}
	cli.PeriodicJobs().Clear()
	cli.PeriodicJobs().Add(river.NewPeriodicJob(
		river.PeriodicInterval(24*time.Hour),
		func() (river.JobArgs, *river.InsertOpts) {
			return CronArgs{}, nil
		},
		nil,
	))
```

**Comment by brandur** [comment, marked as answer]
@assembly-winston Try combining periodic jobs with [unique jobs](https://riverqueue.com/docs/unique-jobs).

Use a unique interval of `24 * time.Hour`, and then have it be inserted periodically to make sure a copy of the job is always inserted.

Notably, don't use a periodic interval of 24 hours because if you did that, your daily job might be inserted considerably late in the day if the start time of the current leader is staggered considerably from midnight. I'd probably inserted the unique job once a minute or once every five minutes, depending on how critical it is that it starts in a very timely manner. That way, on any day rollover a new unique job will be inserted at roughly 12:00 to 12:01 or 12:00 to 12:05 (depending on when the program started) and start running.

**Reply by assembly-winston** [reply]
So I should have the job insertion run in a for loop, and that's the officially recommended usage to have a cron job that I want to run at a specific time once a day, regardless of deploys etc.?

**Reply by bgentry** [reply]
Have you checked out the option to use the cron library? https://riverqueue.com/docs/periodic-jobs#complex-cron-schedules

This covers most use cases, with the exception of the possibility for a missed job if the leader happens to restart right before your scheduled run time.

The advice to combine that with unique jobs lets you be a little more aggressive in trying to insert your daily job any time a new leader takes over, while at the same time ensuring that you can’t actually insert the same job twice in a single day.

**Reply by assembly-winston** [reply]
Hi @bgentry 

I'm doing this now at the bottom of the startup sequence, using the cron library, but it doesn't appear this job runs daily. Am I missing something?

```
func main() {
...
	 schedule, _ := cron.ParseStandard("00 10 * * *")
		cli.PeriodicJobs().Clear()
		cli.PeriodicJobs().Add(river.NewPeriodicJob(
			schedule,
			func() (river.JobArgs, *river.InsertOpts) {
				return CronArgs{}, &river.InsertOpts{UniqueOpts: river.UniqueOpts{
					ByArgs: true,
				}}
			},
			nil,
		))
	}
}
```
